src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,846 @@
+     *     16 (MIC)
+     * </pre>
+     */
+    private static final int NTLM_AUTHENTICATE_PAYLOAD_OFFSET = 88;
+
+    // challenge lengths
+    private static final int NTLM_CLIENT_NONCE_LENGTH = 8;
+    private static final int NTLM_SERVER_CHALLENGE_LENGTH = 8;
+
+    // Windows Filetime timestamp length
+    private static final int NTLM_TIMESTAMP_LENGTH = 8;
+
+    // Windows epoch time difference from Unix epoch time in secs
+    private static final long WINDOWS_EPOCH_DIFF = 11644473600L;
+
+    /**
+     * NTLM Client Context
+     */
+    private class NTLMContext {
+        // domain name to connect to
+        private final String domainName;
+
+        // user credentials
+        private final String upperUserName;
+        private final byte[] passwordHash;
+
+        // unicode bytes
+        private final byte[] domainBytes;
+        private final byte[] userNameBytes;
+        private final byte[] workstationBytes;
+
+        // message authentication code
+        private Mac mac = null;
+
+        // negotiate flags from Challenge msg
+        private long negotiateFlags = 0x00000000;
+
+        // session key calculated from password
+        private byte[] sessionBaseKey = null;
+
+        // Windows FileTime timestamp - number of 100 nanosecond ticks since Windows Epoch time
+        private byte[] timestamp = null;
+
+        // target info field from Challenge msg
+        private byte[] targetInfo = null;
+
+        // server challenge from Challenge msg
+        private byte[] serverChallenge = new byte[NTLM_SERVER_CHALLENGE_LENGTH];
+
+        /**
+         * Section 3.1.5.1.2 Client Receives a CHALLENGE_MESSAGE from the Server
+         * 
+         * Save messages for calculating MIC
+         */
+        private byte[] negotiateMsg = null;
+        private byte[] challengeMsg = null;
+
+        /**
+         * Creates an NTLM client context
+         *
+         * @param domainName
+         * @param userName
+         * @param password
+         * @param workstation
+         * @return NTLM client context
+         * @throws SQLServerException
+         */
+        NTLMContext(final String domainName, final String userName, final String password,
+                final String workstation) throws SQLServerException {
+
+            this.domainName = null != domainName ? domainName.toUpperCase()
+                                                 : SQLServerDriverStringProperty.DOMAIN.getDefaultValue();
+            this.domainBytes = unicode(this.domainName);
+
+            this.userNameBytes = null != SQLServerDriverStringProperty.USER.getDefaultValue() ? unicode(userName)
+                                                                                              : null;
+            this.upperUserName = userName.toUpperCase();
+
+            this.passwordHash = null != password ? MD4(unicode(password)) : null;
+
+            this.workstationBytes = null != workstation ? unicode(workstation.toUpperCase()) : null;

```

Comment:
NTLM specs 2.2.1.1

WorkstationName (variable): If WorkstationLen does not equal 0x0000, WorkstationName MUST be a byte array that contains the name of the client machine that MUST be encoded using the OEM character set. Otherwise, this data is not present.

Reply:
Please also note that when server returns workstation name in Authenticate message, it is in negotiated character set, in our case unicode. 

NTLM specs 2.2.1.3

Workstation (variable): The name of the computer to which the user is logged on. Workstation MUST be encoded in the negotiated character set.